### PR TITLE
Resolving Issue #146

### PR DESCRIPTION
**Pull request by zhongjingjogy**
By add a record item for the length of yticks, the amount of offset  for ylabel is being modified. Two test cases are added, and the results come as following.

test_demo_1

test_demo_2

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -33,4 +34,8 @@ target_include_directories (trase_test PRIVATE tests)
 target_link_libraries (trase_test PRIVATE trase)
 add_test (the_trase_test trase_test)
 
+add_executable(test_trase_ylabel
```
Comment (martinjrobins) [inline tests/CMakeLists.txt]: Tests for this should be in the `TestAxis.cpp` file

File: tests/TestYlabel.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#include "trase.hpp"
+#include <iostream>
+#include <fstream>
+
+using namespace trase;
+
+void test_1() {
```
Comment (martinjrobins) [inline tests/TestYlabel.cpp]: can you please integrate these two tests into the `TestAxis.cpp` file. And add some explicit checks, e.g. check that `m_max_ytick_len` varies between the two cases in the way you expect

File: src/frontend/Axis.tcc [hunk 3]
```diff
@@ -104,10 +108,18 @@ template <typename Backend> void Axis::draw_common_ticks(Backend &backend) {
...
+    if (strlen(buffer) > max_char_len) {
+      max_char_len = strlen(buffer);
+    }
+
     backend.text(vfloat2_t(m_pixels.bmin[0] - m_tick_len / 2, pos), buffer,
                  NULL);
   }
 
+  /// TODO: Convertion from length * font_size to pixels
+  m_max_ytick_len = max_char_len * m_style.font_size() / 96.0 * 72.0;
```
Comment (martinjrobins) [inline src/frontend/Axis.tcc]: Where did 72/96 come from, just trial and error? Yes, agree that we need some way to get the exact pixel length of a string according to the current font, this should be a separate issue

**Review by martinjrobins — CHANGES_REQUESTED** [review]
This is great, thanks @zhongjingjogy. Main thing is to integrate the tests into `TestAxis.cpp` and add some explicit checks

File: src/frontend/Axis.tcc (hunk 3 above)
Comment (zhongjingjogy) [reply src/frontend/Axis.tcc]: There is relation between font and pixels, i.e., for Roboto. It is siad that the scale factor is about 1.25. Anyway, now tests are merged with TestAxis.cpp. However, the current fix comes up with strange "exact pixel length". I am not aware the potential mechanism for the bugs yet.

File: tests/TestAxis.cpp [hunk 4]
```diff
@@ -137,3 +137,82 @@ TEST_CASE("set number of ticks", "[axis]") {
   CHECK(ax->get_ticks()[0] == 10);
   CHECK(ax->get_ticks()[1] == 10);
 }
+
+
+TEST_CASE("ylabel position: case 1", "[axis]") {
+  std::vector<double> predicted_data = {
```
Comment (martinjrobins) [inline tests/TestAxis.cpp:143]: indenting not consistent here.

File: tests/TestAxis.cpp [hunk 5]
```diff
@@ -137,3 +137,82 @@ TEST_CASE("set number of ticks", "[axis]") {
...
+
+    ax->xlabel("Pred. Target");
+    ax->ylabel("Exp. Target");
+
+    // the default number of digits for the labels is set as 2
+    CHECK(ax->max_ytick_len() == int(5 * 18.0 / 1.25));
+
+    // output to svg
+    std::ofstream out;
+    out.open("test_axis_ylabel_position_case1.svg");
```
Comment (martinjrobins) [inline tests/TestAxis.cpp]: use `DummyDraw::draw` like other tests

File: tests/TestAxis.cpp [hunk 6]
```diff
@@ -137,3 +137,82 @@ TEST_CASE("set number of ticks", "[axis]") {
...
+    BackendSVG backend(out);
+    fig->draw(backend);
+    out.close();
+}
+
+TEST_CASE("ylabel position: case 2", "[axis]") {
+  std::vector<double> predicted_data = {
```
Comment (martinjrobins) [inline tests/TestAxis.cpp]: same comments as above for this test

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @zhongjingjogy, just a couple more points on the tests that need tidying up

**Comment by zhongjingjogy** [comment]
I am not aware why the test failed.

```
TEST_CASE("ylabel position: case 1", "[axis]") {
  std::vector<double> predicted_data = {
      -13.8881, -13.8671, -13.8361, -13.7881, -13.7681, -13.7381, -12.6881, -12.661, -12.6481
  };
  std::vector<double> exp_data = {
      -13.8881, -13.8671, -13.8361, -13.7881, -13.7681, -13.7381, -12.6881, -12.661, -12.6481
  };

  auto fig = figure();
  auto ax = fig->axis();
  int predicted_size = predicted_data.size();
  int exp_size = exp_data.size();

  std::vector<float> x(predicted_size);
  std::vector<float> y(exp_size);
  for (int i = 0; i < predicted_size; ++i) {
      x[i] = predicted_data[i];
  }
  for (int i = 0; i < exp_size; ++i) {
      y[i] = exp_data[i];
  }

  auto data = create_data().x(x).y(y);
  auto plt = ax->points(data);

  ax->xlabel("Pred. Target");
  ax->ylabel("Exp. Target");

  // the default number of digits for the labels is set as 2
  std::cout << "ytick char: " << ax->max_ytick_char() << std::endl;
  std::cout << "max_ytick_char_len: " << ax->max_ytick_char_len() << std::endl;
  // NOTE: Nothing is returned.

  CHECK(ax->max_ytick_char_len() == 5);
  CHECK(ax->font_size() == 18.0f);
  CHECK(ax->max_ytick_pixels() == int(ax->max_ytick_char_len() * 18.0 / 1.25));

  DummyDraw::draw("axis", fig);
}
```

The output comes like this,
```
ytick char: 
max_ytick_char_len: 0
```

I would like to kindly ask for some ideas.

**Comment by martinjrobins** [comment]
`Axis::draw_common_ticks` is only set when the axis is drawn (you cannot know the max length of the y ticks before you draw them), but you are checking this value before you draw the figure.

I see the difficulty in this. I don't really want `m_max_ytick_char_len` to be exposed via the public API, as it is easy to be miss-used, but you need to do this to explicitly check the value.

Perhaps don't worry about checking these values explicitly. We need to come up with a better way of testing the visual output of these tests (to check things like overlapping text). Perhaps just snapshot testing, just check that plots havn't changed, and review those that do

So just remove these checks for now, but keep the call to `DummyDraw`.

**Comment by zhongjingjogy** [comment]
That would be great.

**Pull request closed without merging**